frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -5606,6 +5495,312 @@ const Decl* findFieldByName(Context* context,
+    CHPL_ASSERT(iter->isPromotionIteratorType());
+    id = iter->toPromotionIteratorType()->scalarFn()->id();
+  }
+  auto iteratorScope = scopeForId(context, id);
+  auto callScope = iteratorScope;
+  auto poiScope = iter->poiScope();
+
+  // If the function needs a PoI scope, this scope will capture functions
+  // at the iterator's own point of instantiation; we don't want to include
+  // this scope in the overload search (lookup scope), because that would make
+  // it possible to introduce new overloads of the iterator via PoI, which we do not
+  // want to allow.
+  //
+  // However, we do want to include the functions available at instantiation time
+  // when resolving the bodies of the other overloads, if applicable. So,
+  // change to callScope to unwrap one level from the PoI scope.
+  //
+  // Loop expressions do not create new PoI scopes (should they?) so they
+  // are exempt from this.
+  //
+  // See the comment on CallScopeInfo for details on why three scopes are
+  // necessary for resolving functions.
+  if (poiScope && !iter->isLoopExprIteratorType()) {
+    callScope = poiScope->inScope();
+    poiScope = poiScope->inFnPoi();
+  }
+
+  return CallScopeInfo::forIteratorOverloadSearch(callScope, iteratorScope, poiScope);
+}
+
+const MostSpecificCandidate&
+findTaggedIteratorForType(ResolutionContext* rc,
+                          const FnIteratorType* fnIter,
+                          Function::IteratorKind iterKind) {
+  CHPL_RESOLUTION_QUERY_BEGIN(findTaggedIteratorForType, rc, fnIter, iterKind);
+
+  auto name = fnIter->iteratorFn()->untyped()->name();
+  auto receiverType =
+    fnIter->iteratorFn()->isMethod() ?
+    fnIter->iteratorFn()->formalType(0) :
+    QualifiedType();
+  std::vector<QualifiedType> argTypes;
+  for (int i = 0; i < fnIter->iteratorFn()->numFormals(); i++) {
+    argTypes.push_back(fnIter->iteratorFn()->formalType(i));
+  }
+  auto inScopes = callScopeInfoForIterator(rc->context(), fnIter);
+
+  auto ret = findTaggedIterator(rc, name, receiverType, argTypes, iterKind,
+                                inScopes.callScope(), inScopes.lookupScope(), inScopes.poiScope());
+  return CHPL_RESOLUTION_QUERY_END(ret);
+}
+
+const types::QualifiedType&
+taggedYieldTypeForType(ResolutionContext* rc,
+                       const types::FnIteratorType* fnIter,
+                       uast::Function::IteratorKind iterKind) {
+  CHPL_RESOLUTION_QUERY_BEGIN(taggedYieldTypeForType, rc, fnIter, iterKind);
+
+  // building a call resolution result takes care of setting up the
+  // PoI scope and doing type inference using that PoI scope, so do it here.
+  auto msc = findTaggedIteratorForType(rc, fnIter, iterKind);
+  auto inScopes = callScopeInfoForIterator(rc->context(), fnIter);
+  auto c = resolutionResultFromMostSpecificCandidate(rc, msc, inScopes);
+
+  QualifiedType ret;
+  if (!c.exprType().isUnknownOrErroneous()) {
+    if (auto it = c.exprType().type()->toIteratorType()) {
+      ret = it->yieldType();
```

Comment:
Nothing is necessarily wrong but I'm missing something. When is the IteratorType including the yield type computed? I thought that here we were going to be resolving the iterator `Function` through normal means (and so we would need to separately compute the yield/return type after we have found the most specific candidate TypedFnSignature).

Reply:
You're right, storing it is not needed. I've made [some changes](https://github.com/chapel-lang/chapel/pull/26039/files/7233c8e212727d25ded6ea8fbf54f96ce07eded1..a8e62dd1a6beec6caec1660ad867635f1d5e678f) to only store `yieldType` for loops, and use queries / stored types for iterator fns.